core/jvm/src/main/scala/scalaz/zio/blocking/package.scala

```diff
@@ -18,7 +18,14 @@ package scalaz.zio
 
 import scalaz.zio.internal.Executor
 
-package object blocking extends Blocking.Service[Blocking] {
+package object blocking {
+  final def blockingExecutor: ZIO[Blocking, Nothing, Executor]                = blockingPackage.blockingExecutor
+  final def blocking[R1 <: Blocking, E, A](zio: ZIO[R1, E, A]): ZIO[R1, E, A] = blockingPackage.blocking(zio)
+  final def interruptible[A](effect: => A): ZIO[Blocking, Throwable, A]       = blockingPackage.interruptible(effect)
+}
+
+import blocking._
+private object blockingPackage extends Blocking.Service[Blocking] {
```

Comment:
If you turn `blocking` into an `object` instead of a `package object`, but still let it extend `Blocking.Service[Blocking]`, will the code work?

That would be simpler. We don't need `blocking` to be a package object, it can be an ordinary object instead.

Reply:
That would have been nice, but it doesn't work due to the name collision with the package `blocking`.
If we change the name, it would work fine!
```
[error] /Users/jon/Project/scalaz-zio/core/jvm/src/main/scala/scalaz/zio/blocking/package.scala:22:8: blocking is already defined as package blocking
[error] object blocking extends Blocking.Service[Blocking] {
[error]        ^
[error] one error found
```

Reply:
What if we delete the package and move its contents into the object???

The drawback with the above encoding is that there is no guarantee if someone adds a new method to `Blocking.Service`, that the compiler makes them add the same method to the package object. So the whole `extends Blocking.Service[Blocking]` thing isn't useful. In fact, we may as well delete that, in which case the package object will probably work just fine, because it won't be extending that trait.

So ideally we can find a way to make the object extend the trait.

Reply:
Tried already, the package is the package of other classes and traits like `blocking`
```
package scalaz.zio.blocking
trait Blocking extends Serializable {
  val blocking: Blocking.Service[Any]
}
```

That being said, I agree with you. The technique used here helps a bit, it allows getting a definition for things that are already defined in the trait like `blocking. interruptible`.

A variation of it would be to have:
```
package object blocking {
      //

   object self extends Blocking.Service[Blocking] {
      //
   }
}
```

Reply:
Sure, I meant to move `Blocking` (the trait) into the `blocking` object, and literally delete the package. I am not sure if it would be legal to do: `object blocking extends blocking.Blocking.Service[blocking.Blocking`, but it could be worth a try. 

Failing that, I'd vote for, I guess, deleting the private objects and just manually implementing the methods inside each `object`, taking care to implement all of them. In this case, we don't need the `R` type parameter in `Service` because it exists only to facilitate this.

Reply:
```
[error] /Users/jon/Project/scalaz-zio/core/jvm/src/main/scala/scalaz/zio/blocking/package.scala:21:25: illegal cyclic reference involving object blocking
```

That style doesn't help much: https://gist.github.com/ahoy-jon/827e585314314b3f4925468de4f97d30

The problem I find with that style is duplication of existing definitions from `Blocking.Service`: 
https://gist.github.com/ahoy-jon/f96c3ee74428c1f2ba3a5cb4920573c6

I am trying something nice in that direction.

Reply:
Yes, if we are duplicating, we may as well do that in the package object. 

The 2nd gist does not work since all services in ZIO must be testable / mockable. So we can't hard code the "helper functions" in the package object to any specific implementation of the service definition.

Overall, I'm leaning toward just having the package object, with `N` functions that all access the service through `accessM`; the function signatures and types should be the same as those inside the `Service` trait, even though there doesn't appear a way to enforce that in Scala 2.11.

Reply:
Tried a bit : 
https://gist.github.com/ahoy-jon/280d1d91eb682d7cf597371e09507e04 

The idea is to avoid duplicate implementations of `interruptible`.

Reply:
Ok, just like that : 
```scala

package object blocking {
  final def blockingExecutor: ZIO[Blocking, Nothing, Executor]                = ZIO.accessM(_.blocking.blockingExecutor)
  final def blocking[R1 <: Blocking, E, A](zio: ZIO[R1, E, A]): ZIO[R1, E, A] = ZIO.accessM(_.blocking.blocking(zio))
  final def interruptible[A](effect: => A): ZIO[Blocking, Throwable, A]       = ZIO.accessM(_.blocking.interruptible(effect))
}
```

Reply:
Yes, that should work fine. It just means we need to "manually" keep the helper methods in sync with the service methods.